plugins/repository-s3/src/test/java/org/elasticsearch/repositories/s3/S3BlobContainerRetriesTests.java

```diff
@@ -0,0 +1,385 @@
+                    return;
+                }
+
+            } else if ("POST".equals(exchange.getRequestMethod())
+                        && exchange.getRequestURI().getQuery().equals("uploadId=TEST")) {
+                // complete multipart upload request
+                Streams.readFully(exchange.getRequestBody());
+                if (countDownComplete.countDown()) {
+                    byte[] response = ("<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n" +
+                        "<CompleteMultipartUploadResult>\n" +
+                        "  <Bucket>bucket</Bucket>\n" +
+                        "  <Key>write_large_blob</Key>\n" +
+                        "</CompleteMultipartUploadResult>").getBytes(StandardCharsets.UTF_8);
+                    exchange.getResponseHeaders().add("Content-Type", "application/xml");
+                    exchange.sendResponseHeaders(HttpStatus.SC_OK, response.length);
+                    exchange.getResponseBody().write(response);
+                    exchange.close();
+                    return;
+                }
+            }
+
+            // sends an error back or let the request time out
+            if (useTimeout == false) {
+                exchange.sendResponseHeaders(randomFrom(HttpStatus.SC_INTERNAL_SERVER_ERROR, HttpStatus.SC_BAD_GATEWAY,
+                                                        HttpStatus.SC_SERVICE_UNAVAILABLE, HttpStatus.SC_GATEWAY_TIMEOUT), -1);
+                exchange.close();
+            }
+        });
+
+        blobContainer.writeBlob("write_large_blob", new ZeroInputStream(blobSize), blobSize, false);
+
+        assertThat(countDownInitiate.isCountedDown(), is(true));
+        assertThat(countDownUploads.get(), equalTo(0));
+        assertThat(countDownComplete.isCountedDown(), is(true));
+    }
+
+    /**
+     * A resettable InputStream that only serves zeros.
+     *
+     * Ideally it should be wrapped into a BufferedInputStream but it seems that the AWS SDK is calling InputStream{@link #reset()}
+     * before calling InputStream{@link #mark(int)}, which is not permitted by the {@link #reset()} method contract.
+     **/
+    private static class ZeroInputStream extends InputStream {
+
+        private final AtomicBoolean closed = new AtomicBoolean(false);
+        private final long length;
+        private final AtomicLong reads;
+        private volatile long mark;
+
+        private ZeroInputStream(final long length) {
+            this.length = length;
+            this.reads = new AtomicLong(length);
+            this.mark = -1;
+        }
+
+        @Override
+        public int read() throws IOException {
+            ensureOpen();
+            if (reads.decrementAndGet() < 0) {
+                return -1;
+            }
+            return 0;
+        }
+
+        @Override
+        public boolean markSupported() {
+            return true;
+        }
+
+        @Override
+        public synchronized void mark(int readlimit) {
```

Comment:
we're relying on the fact that the implementation is ignoring the read limit, as the AWS SDK sets a low limit here (~131kb).

The mark/reset behavior of the SDK is also dependent on whether we have a `BufferedInputStream` (as you pointed out).
It would be good to test the behavior of the actual inputstream that we are passing in.

In particular, the assertion added by #45224 is not sufficient to guarantee us retry behavior.

Reply:
> we're relying on the fact that the implementation is ignoring the read limit, as the AWS SDK sets a low limit here (~131kb).

As far as I understand the InputStream documentation, it seems that the read limit might or might not be respected by implementations. In any case the AWS SDK provides a default read limit (131kb) when marking the input stream but does not really expect the stream to respect the limit. The only exception is when the provided input stream is a `BufferedInputStream` and in this case the request client options read limit must be configured accordingly.

I've been confused by this because at first I tried to use a `BufferedInputStream` in the test class but the AWS SDK has some special treatment for such streams, and since we don't provide buffered input stream to the client when snapshotting data I think it just does not make sense to use it here.

 I opened #45847 to beef up the tests a bit so that we sometimes upload blobs that are large than the default read limit and that sometimes we consumes part of the request body. 

>  It would be good to test the behavior of the actual inputstream that we are passing in.

I agree, I'll add such tests as a follow up.